File: minimum-spanning-trees/python/kruskal.py
from graph import Edge, Graph, UnionFind, Vertex


def kruskal(graph):
    vertices = graph.vertices
    edges = graph.edges
    union_find = UnionFind("".join(vertices.keys()))

    edges.sort()

    results = []
    for edge in edges:
        start = edge.start
        end = edge.end
        weight = edge.weight
        if union_find.find_set(start) != union_find.find_set(end):
            print(f"append edge: start = {start}, end = {end}, weight = {weight}")

            results.append(edge)

            union_find.union(start, end)

    return results


if __name__ == "__main__":
    vertices = {}
    for name in "abcdefghi":
        vertices[name] = Vertex(name)

    edges = []
    edges.append(Edge("a", "b", 4))
    edges.append(Edge("a", "h", 8))
    edges.append(Edge("b", "a", 4))
    edges.append(Edge("b", "c", 8))
    edges.append(Edge("b", "h", 11))
    edges.append(Edge("c", "b", 8))
    edges.append(Edge("c", "d", 7))
    edges.append(Edge("c", "f", 4))
    edges.append(Edge("c", "i", 2))
    edges.append(Edge("d", "c", 7))
    edges.append(Edge("d", "e", 9))
    edges.append(Edge("d", "f", 14))
    edges.append(Edge("e", "d", 9))
    edges.append(Edge("e", "f", 10))
    edges.append(Edge("f", "c", 4))
    edges.append(Edge("f", "d", 14))
    edges.append(Edge("f", "e", 10))
    edges.append(Edge("f", "g", 2))
    edges.append(Edge("g", "f", 2))
    edges.append(Edge("g", "h", 1))
    edges.append(Edge("g", "i", 6))
    edges.append(Edge("h", "a", 8))
    edges.append(Edge("h", "b", 11))
    edges.append(Edge("h", "g", 1))
    edges.append(Edge("h", "i", 7))
    edges.append(Edge("i", "c", 2))
    edges.append(Edge("i", "g", 6))
    edges.append(Edge("i", "h", 7))

    graph = Graph(vertices, edges)

    print("--- graph ---")
    graph.print_graph()

    print("--- Kruskal's algorithm ---")
    results = kruskal(graph)

    print("--- result ---")
    total = 0
    for edge in results:
        print(f"start = {edge.start}, end = {edge.end}, weight = {edge.weight}")

        total += edge.weight

    print(f"total = {total}")

File: minimum-spanning-trees/golang/kruskal.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

func krsukal(graph *Graph) []*Edge {
	edges := graph.edges

	sort.Slice(edges, func(i, j int) bool {
		return edges[i].weight < edges[j].weight
	})

	var names []string
	for name, _ := range graph.vertices {
		names = append(names, name)
	}

	unionFind := &UnionFind{}
	unionFind.initializeUnionSet(strings.Join(names, ""))

	var results []*Edge
	for _, edge := range edges {
		start := edge.start
		end := edge.end
		if unionFind.findSet(start) != unionFind.findSet(end) {
			fmt.Printf("append edge: start = %s, end = %s, weight = %d\n", start, end, edge.weight)
			results = append(results, edge)

			unionFind.union(start, end)
		}
	}

	return results
}

func main() {
	vertices := make(map[string]*Vertex)
	for _, name := range "abcdefghi" {
		vertices[string(name)] = &Vertex{name: string(name)}
	}

	edges := make([]*Edge, 0)
	edges = append(edges, &Edge{start: "a", end: "b", weight: 4})
	edges = append(edges, &Edge{start: "a", end: "h", weight: 8})
	edges = append(edges, &Edge{start: "b", end: "a", weight: 4})
	edges = append(edges, &Edge{start: "b", end: "c", weight: 8})
	edges = append(edges, &Edge{start: "b", end: "h", weight: 11})
	edges = append(edges, &Edge{start: "c", end: "b", weight: 8})
	edges = append(edges, &Edge{start: "c", end: "d", weight: 7})
	edges = append(edges, &Edge{start: "c", end: "f", weight: 4})
	edges = append(edges, &Edge{start: "c", end: "i", weight: 2})
	edges = append(edges, &Edge{start: "d", end: "c", weight: 7})
	edges = append(edges, &Edge{start: "d", end: "e", weight: 9})
	edges = append(edges, &Edge{start: "d", end: "f", weight: 14})
	edges = append(edges, &Edge{start: "e", end: "d", weight: 9})
	edges = append(edges, &Edge{start: "e", end: "f", weight: 10})
	edges = append(edges, &Edge{start: "f", end: "c", weight: 4})
	edges = append(edges, &Edge{start: "f", end: "d", weight: 14})
	edges = append(edges, &Edge{start: "f", end: "e", weight: 10})
	edges = append(edges, &Edge{start: "f", end: "g", weight: 2})
	edges = append(edges, &Edge{start: "g", end: "f", weight: 2})
	edges = append(edges, &Edge{start: "g", end: "h", weight: 1})
	edges = append(edges, &Edge{start: "g", end: "i", weight: 6})
	edges = append(edges, &Edge{start: "h", end: "a", weight: 8})
	edges = append(edges, &Edge{start: "h", end: "b", weight: 11})
	edges = append(edges, &Edge{start: "h", end: "g", weight: 1})
	edges = append(edges, &Edge{start: "h", end: "i", weight: 7})
	edges = append(edges, &Edge{start: "i", end: "c", weight: 2})
	edges = append(edges, &Edge{start: "i", end: "g", weight: 6})
	edges = append(edges, &Edge{start: "i", end: "h", weight: 7})

	graph := &Graph{vertices: vertices, edges: edges}
	graph.initializeGraph()

	fmt.Println("--- graph ---")
	graph.printGraph()

	fmt.Println("--- Kruskal's algorithm ---")
	results := krsukal(graph)

	fmt.Println("--- result ---")
	total := 0
	for _, edge := range results {
		fmt.Printf("start = %s, end = %s, weight = %d\n", edge.start, edge.end, edge.weight)

		total += edge.weight
	}
	fmt.Printf("total = %d\n", total)
}
